### PR TITLE
Merging changes up to this point

### PR DESCRIPTION
**Pull request by pleblanc1976**

### CODE REVIEW

**Comment by pleblanc1976** [comment]
The failover agent isn't tested yet.  I hope to get to that tomorrow.

File: cmapi_server/cmapi_server.conf [hunk 1]
```diff
@@ -7,6 +7,7 @@ server.ssl_private_key = './cmapi_server/self-signed.key'
 engine.autoreload.on = False
 log.access_file = ''
 log.error_file = ''
+api_version = '0.3.0'
```
Comment (drrtuy) [inline cmapi_server/cmapi_server.conf:10]: Questionable change that promotes bash scripts usage that I don't like. Why can't we put it as a python object in the code? 
Another philosophical question: what happens with the server if a user decides to reduce the version in the config? Will the config change bring the functionality of version the user sets?

File: cmapi_server/controllers/endpoints.py [hunk 2]
```diff
@@ -209,6 +210,9 @@ def put_config(self):
         request_revision = request_body.get('revision', None)
         request_manager = request_body.get('manager', None)
         request_timeout = request_body.get('timeout', None)
+        
+        # is_test = True means this should not save the config file or apply the changes
+        is_test = request_body.get('test', False)
```
Comment (drrtuy) [inline cmapi_server/controllers/endpoints.py:215]: I had them some time ago but this extends the swagger spec w/o documenting this. I decided to remove all these tests. If you thing it is good to revive it then we need to update the API spec.

**Comment by drrtuy** [comment]
I wonder if the changes passed the existing unit tests?

File: cmapi_server/failover_agent.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+'''
+
+import requests
+from failover.agent_comm import AgentBase
+import helpers
+import logging
+import node_manipulation
+from mcs_node_control.node_config import NodeConfig
+
+logger = logging.getLogger()
```
Comment (drrtuy) [inline cmapi_server/failover_agent.py:13]: I would suggest to export Logger settings using examples from mcs_node_controller/controller.

File: cmapi_server/cmapi_server.conf (hunk 1 above)
Comment (pleblanc1976) [reply cmapi_server/cmapi_server.conf:10]: I added that b/c last week you said the version should go in the config file.  :D  

We can do whatever we want with it though.  Putting into a python object would be fine.  Config file is also fine IMO.  A user can always cause all kinds of trouble through config files; to some extent they have to be trusted to not break their own system.

File: cmapi_server/controllers/endpoints.py (hunk 2 above)
Comment (pleblanc1976) [reply cmapi_server/controllers/endpoints.py:215]: Having the option to not modify the system was very useful in testing my own code, so I think we should keep this for now.  I don't think we need to publish docs for it; it's a var for our usage only, not for general use.

File: cmapi_server/helpers.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                r.raise_for_status()
+            except requests.Timeout as e:
+                logger.warning(f"commit_transaction(): timeout on node {node}")
+            except Exception as e:
+                logger.warning(f"commit_transaction(): got error during request to {node}: {str(e)}")
+            else:
+                break
+            time.sleep(1)
+            
+def broadcast_new_config(cs_config_filename="/etc/columnstore/Columnstore.xml", cmapi_config_filename = "/etc/columnstore/cmapi_server.conf", test_mode = False):
```
Comment (drrtuy) [inline cmapi_server/helpers.py:132]: We have this hardcoded somewhere already.

Comment (drrtuy) [reply cmapi_server/helpers.py:132]: Default SM config path should go into the same place.

File: cmapi_server/helpers.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    version = get_version(cfg_parser)
+
+    nc = NodeConfig()
+    root = nc.get_current_config_root(config_filename = cs_config_filename)
+    with open(cs_config_filename) as f:
+        config_text = f.read()
+
+    headers = {"x-api-key": key}
+    body = {
+        "manager": root.find("./ClusterManager").text,
```
Comment (drrtuy) [inline cmapi_server/helpers.py:145]: Shouldn't we increase both counters?

File: cmapi_server/helpers.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    if test_mode:
+        body["test"] = True
+
+    active_nodes = get_active_nodes()
+    failed_nodes = []
+    success_nodes = []
+    for node in active_nodes:
+        url = f"https://{node}:8640/cmapi/{version}/node/config"
+        success = False
+        for retry in range(5):
```
Comment (drrtuy) [inline cmapi_server/helpers.py:159]: This limit should be configurable.

File: cmapi_server/helpers.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        if not success:
+            failed_nodes.append(node)
+    
+    if len(success_nodes) > 0:
+        logger.info(f"broadcast_new_config(): successfully pushed the new config file to {success_nodes}")
+    if len(failed_nodes) > 0:
+        logger.error(f"broadcast_new_config(): failed to push the new config file to {failed_nodes}")
+        return False
+    return True
+    # Hm, should we raise an exception, and under what conditions...
```
Comment (drrtuy) [inline cmapi_server/helpers.py:181]: It depends on where are we using this. Would it be a REST server then no.

File: cmapi_server/test/CS-config-test.xml [hunk 8]
```diff
@@ -0,0 +1,563 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (drrtuy) [inline cmapi_server/test/CS-config-test.xml:1]: I would suggest to remove this header and put the comments inside Columnstore tag b/c their placement violates the spec IIRC.

File: cmapi_server/failover_agent.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            except Exception as e:
+                logger.error(f"FA.deactivateNodes(): failed to deactivate node {node}, got {str(e)}")
+                
+        if old_node_count > 0:
+            try:
+                helpers.update_revision_and_manager()
+                helpers.broadcast_new_config()
+            except Exception as e:
+                logger.error(f"FA.deactivateNodes(): failed to broadcast the new configuration, got {str(e)}, rolling back")
+                helpers.rollback_transaction(txn_id)
```
Comment (drrtuy) [inline cmapi_server/failover_agent.py:71]: Here might happen another exception.

File: cmapi_server/helpers.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+import datetime
+import time
+
+from mcs_node_control.models.node_config import NodeConfig
+
+logger = logging.getLogger()
+
+''' returns (bool, txnid).  Bool = True means it successfully started a transaction.
+    False means it didn't.  If True, then txnid will have the transaction ID '''
+def start_transaction(config_filename = "/etc/columnstore/cmapi_server.conf"):
```
Comment (drrtuy) [inline cmapi_server/helpers.py:15]: Point the hardcode to a common object.

File: cmapi_server/helpers.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    cfg_parser = get_config_parser(config_filename = config_filename)
+    api_key = get_current_key(cfg_parser)
+    id = get_id()
+    version = get_version(cfg_parser)
+    
+    headers = {'x-api-key': api_key}
+    body = {'id' : id}
```
Comment (drrtuy) [inline cmapi_server/helpers.py:22]: This call must fail w/o a timeout.

File: cmapi_server/failover_agent.py (hunk 9 above)
Comment (pleblanc1976) [reply cmapi_server/failover_agent.py:72]: there are definitely more todo items around error handling.  It's not bulletproof yet.

**Pull request merged**